In [95]:
import pandas as pd
pd.set_option("display.max_columns", 101)

# Clean and prepare dataset

We create a dataframe named df_pivot that contains job features (skills and knowledge) as columns and job titles as rows. For the purpose of this project, jobs are filtered depending on their job zones. 
All the data can be found here:
https://www.onetcenter.org/database.html

In [96]:
# Load datasets into Pandas DataFrame
skills = pd.read_excel('../databases/ONET/Skills.xlsx')

In [97]:
knowledge = pd.read_excel('../databases/ONET/Knowledge.xlsx')

In [98]:
# Merge skills and knowledge files
df = skills.append(knowledge)

In [99]:
# Filter per job zones 3, 4 and 5
job_zones = pd.read_excel('../databases/ONET/Job Zones.xlsx')
jobzone345 = job_zones[job_zones['Job Zone'].isin([3,4,5])]['O*NET-SOC Code'].tolist()
df = df[df['O*NET-SOC Code'].isin(jobzone345)]

In [100]:
# Take out unnecessary columns
cols_to_keep = [
    'O*NET-SOC Code', 
    'Title', 
    'Element Name', 
    'Scale ID',
    'Data Value'
]

df = df[cols_to_keep].copy()

In [101]:
df.head()

O*NET-SOC Code             Title           Element Name Scale ID  Data Value
0     11-1011.00  Chief Executives  Reading Comprehension       IM        4.12
1     11-1011.00  Chief Executives  Reading Comprehension       LV        4.75
2     11-1011.00  Chief Executives       Active Listening       IM        4.12
3     11-1011.00  Chief Executives       Active Listening       LV        4.88
4     11-1011.00  Chief Executives                Writing       IM        4.00

In [102]:
# Merge columns 'Element Name' and 'Scale ID' into 'element_scale_id'
df['element_scale_id'] = df['Element Name'] + '_' + df['Scale ID'] 

In [103]:
# Pivot table
df_pivot = df.pivot_table(index=['Title'], columns='element_scale_id', values='Data Value')

In [104]:
df_pivot.head()

element_scale_id                        Active Learning_IM  \
Title                                                        
Accountants                                           3.25   
Actuaries                                             3.38   
Acupuncturists                                        3.12   
Acute Care Nurses                                     3.75   
Adapted Physical Education Specialists                3.50   

element_scale_id                        Active Learning_LV  \
Title                                                        
Accountants                                           3.38   
Actuaries                                             4.12   
Acupuncturists                                        3.50   
Acute Care Nurses                                     3.88   
Adapted Physical Education Specialists                3.62   

element_scale_id                        Active Listening_IM  \
Title                                                         
Accountants                                            3.88   
Actuaries                                              4.00   
Acupuncturists                                         3.75   
Acute Care Nurses                                      4.00   
Adapted Physical Education Specialists                 4.12   

element_scale_id                        Active Listening_LV  \
Title                                                         
Accountants                                            3.62   
Actuaries                                              4.12   
Acupuncturists                                         3.62   
Acute Care Nurses                                      3.88   
Adapted Physical Education Specialists                 4.12   

element_scale_id                        Administration and Management_IM  \
Title                                                                      
Accountants                                                         3.43   
Actuaries                                                           3.31   
Acupuncturists                                                      2.67   
Acute Care Nurses                                                   2.89   
Adapted Physical Education Specialists                              2.44   

element_scale_id                        Administration and Management_LV  \
Title                                                                      
Accountants                                                         4.00   
Actuaries                                                           4.72   
Acupuncturists                                                      2.57   
Acute Care Nurses                                                   2.96   
Adapted Physical Education Specialists                              2.56   

element_scale_id                        Biology_IM  Biology_LV  \
Title                                                            
Accountants                                   1.02        0.03   
Actuaries                                     1.16        0.47   
Acupuncturists                                3.45        4.05   
Acute Care Nurses                             3.32        3.50   
Adapted Physical Education Specialists        2.65        2.81   

element_scale_id                        Building and Construction_IM  \
Title                                                                  
Accountants                                                     1.21   
Actuaries                                                       1.25   
Acupuncturists                                                  1.60   
Acute Care Nurses                                               1.18   
Adapted Physical Education Specialists                          1.58   

element_scale_id                        Building and Construction_LV  \
Title                                                                  
Accountants                                                     0.40

In [105]:
# Display size of the matrix
df_pivot.shape

(640, 134)

Let's factorize this matrix to (640, 2)

# 2D projection with PCA

We identify the 2 principal components with the highest variance with PCA, and project each point on the resulting 2D space.

In [106]:
# Set up
import numpy as np
from sklearn.decomposition import PCA

In [107]:
pca = PCA(n_components=2)

In [108]:
pc  = pca.fit_transform(df_pivot)

In [112]:
pc

array([[-3.34835972,  5.18261815],
       [-3.56159461,  1.22851155],
       [-5.78468294, -1.48326019],
       ...,
       [ 2.71973696, -1.89570263],
       [ 9.36660871,  0.86357522],
       [-0.19715302, -3.09207539]])

In [113]:
pc.shape

(640, 2)

We now have an array of floats that correspond to the coordinates of each point of the dataset in the target space. Let's visualize it.

# Visualization with plotly

In [110]:
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected=True)
import plotly.graph_objs as go

In [124]:
data = [
    go.Scatter(
        x=pc[:,0],
        y=pc[:,1],
        mode='markers',
        text=df_pivot.index.tolist(),
        opacity=0.5
    )
]
layout = go.Layout(
    title='Jobs similarity map (based on skills and knowledge) with PCA',
    hovermode='closest'
)
fig = go.Figure(data=data, layout=layout)
plot_url = iplot(fig, filename='hover-chart-basic')